gateway-js/src/index.ts

```diff
@@ -545,28 +596,81 @@ export class ApolloGateway implements GraphQLService {
+  // again. Note that it is an async function whose Promise is not actually awaited;
+  // it should never throw itself other than due to a bug in its state machine.
+  private async pollServices(unrefTimer: boolean) {
+    switch (this.state.phase) {
+      case 'stopping':
+      case 'stopped':
+        return;
+      case 'initialized':
+        throw Error('pollServices should not be called before load!');
+      case 'polling':
+        throw Error(
+          'pollServices should not be called while in the middle of polling!',
+        );
+      case 'waiting to poll':
+        throw Error(
+          'pollServices should not be called while already waiting to poll!',
+        );
+      case 'loaded':
+        // This is the normal case.
+        break;
+      default:
+        throw new UnreachableCaseError(this.state);
+    }
 
-    // Sleep for the specified pollInterval before kicking off another round of polling
-    await new Promise<void>((res) => {
-      this.pollingTimer = setTimeout(
-        () => res(),
-        this.experimental_pollInterval || 10000,
-      );
-      // Prevent the Node.js event loop from remaining active (and preventing,
-      // e.g. process shutdown) by calling `unref` on the `Timeout`.  For more
-      // information, see https://nodejs.org/api/timers.html#timers_timeout_unref.
-      this.pollingTimer?.unref();
+    // Transition into 'waiting to poll' and set a timer. The timer resolves the
+    // Promise we're awaiting here; note that calling stop() also can resolve
+    // that Promise.
+    await new Promise<void>((doneWaiting) => {
+      this.state = {
+        phase: 'waiting to poll',
+        doneWaiting,
+        pollWaitTimer: setTimeout(() => {
+          // Note that we might be in 'stopped', in which case we just do
+          // nothing.
+          if (this.state.phase == 'waiting to poll') {
+            this.state.doneWaiting();
+          }
+        }, this.experimental_pollInterval || 10000),
+      };
+      if (unrefTimer) {
+        this.state.pollWaitTimer.unref();
+      }
     });
 
+    // If we've been stopped, then we're done. The cast here is because TS
+    // doesn't understand that this.state can change during the await
+    // (https://github.com/microsoft/TypeScript/issues/9998).
+    if ((this.state as GatewayState).phase !== 'waiting to poll') {
+      return;
+    }
+
+    let pollingDone: () => void;
+    this.state = {
+      phase: 'polling',
+      pollingDonePromise: new Promise<void>((res) => {
+        pollingDone = res;
+      }),
+    };
+
     try {
       await this.updateComposition();
     } catch (err) {
       this.logger.error((err && err.message) || err);
     }
 
-    this.pollServices();
+    if (this.state.phase === 'polling') {
+      // If we weren't stopped, we should transition back to the initial 'loading' state and trigger
+      // another call to itself. (Do that in a setImmediate to avoid unbounded stack sizes.)
```

Comment:
Cool! TIL

Reply:
FWIW I actually don't know offhand if you get unbounded stack sizes from recursion in async functions.

Reply:
(I checked and you do.)